gax-java/gax-grpc/src/main/java/com/google/api/gax/grpc/InstantiatingGrpcChannelProvider.java

```diff
@@ -345,13 +333,29 @@ boolean isCredentialDirectPathCompatible() {
   static boolean isOnComputeEngine() {
     String osName = System.getProperty("os.name");
     if ("Linux".equals(osName)) {
+      String systemProductName = getSystemProductName();
```

Comment:
IIUC, this should be assigned to back static instance field, right? The value would cached and we shouldn't need to re-read the file on future channel creations.

```suggestion
     systemProductName = getSystemProductName();
```

Perhaps the systemProductName should also be volatile since it *may* (?) be possible that we may re-size a channel while the `isOnComputeEngine` runs?